exp/internal/controllers/machinepool_controller_phases.go

```diff
@@ -307,3 +315,235 @@ func (r *MachinePoolReconciler) reconcileInfrastructure(ctx context.Context, clu
 
 	return ctrl.Result{}, nil
 }
+
+// reconcileMachines reconciles Machines associated with a MachinePool.
+//
+// Note: In the case of MachinePools the machines are created in order to surface in CAPI what exists in the
+// infrastructure while instead on MachineDeployments, machines are created in CAPI first and then the
+// infrastructure is created accordingly.
+// Note: When supported by the cloud provider implementation of the MachinePool, machines will provide a means to interact
+// with the corresponding infrastructure (e.g. delete a specific machine in case MachineHealthCheck detects it is unhealthy).
+func (r *MachinePoolReconciler) reconcileMachines(ctx context.Context, mp *expv1.MachinePool, infraMachinePool *unstructured.Unstructured) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	var noKind bool
+	var noSelector bool
+
+	var infraMachineKind string
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineKind, "status", "infrastructureMachineKind"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noKind = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineKind from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+
+	var infraMachineSelector metav1.LabelSelector
+	if err := util.UnstructuredUnmarshalField(infraMachinePool, &infraMachineSelector, "status", "infrastructureMachineSelector"); err != nil {
+		if errors.Is(err, util.ErrUnstructuredFieldNotFound) {
+			noSelector = true
+		} else {
+			return errors.Wrapf(err, "failed to retrieve infraMachineSelector from infrastructure provider for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
+		}
+	}
+
+	// If one missing but not both, return error
+	if noKind != noSelector {
+		return errors.Errorf("only one of infraMachineKind or infraMachineSelector found, both must be present")
+	}
+
+	// If MachinePool Machines aren't supported by the InfraMachinePool, skip reconcile and return early.
+	if noKind && noSelector {
+		log.V(2).Info("MachinePool Machines not supported, no infraMachineKind or infraMachineSelector found")
+		return nil
+	}
+
+	log.Info("Reconciling MachinePool Machines", "infrastructureMachineKind", infraMachineKind, "infrastructureMachineSelector", infraMachineSelector)
+	var infraMachineList unstructured.UnstructuredList
+
+	infraMachineList.SetAPIVersion(infraMachinePool.GetAPIVersion())
+	infraMachineList.SetKind(infraMachineKind)
```

Comment:
Let's please add the suffix here. It technically works without because CR trims it at some point (which is a no-op if you omit it). But I think it's more correct to actually request the kind that we want and not depend on this CR behavior

Please also move this down to the list call, like this:
```go
	var infraMachineList unstructured.UnstructuredList
	infraMachineList.SetAPIVersion(infraMachinePool.GetAPIVersion())
	infraMachineList.SetKind(infraMachineKind)
	if err := r.Client.List(ctx, &infraMachineList, client.InNamespace(mp.Namespace), client.MatchingLabels(infraMachineSelector.MatchLabels)); err != nil {
		return errors.Wrapf(err, "failed to list infra machines for MachinePool %q in namespace %q", mp.Name, mp.Namespace)
	}
```
It's just a lot easier to read if we don't mix the definitions of output parameters of various funcs with their calls (Watch & List in this case)

Reply:
Would we just want to change this to `infraMachineList.SetKind(infraMachineKind + "List")`?

Reply:
Yes!